### PR TITLE
feat: add og

### PR DESCRIPTION
**Pull request by Innei**
### <samp>🤖 Generated by Copilot at 92c5aeb</samp>

This pull request adds a feature to generate dynamic Open Graph images for note pages using an edge function. It modifies the `page.tsx` component to use the `api/og` endpoint when the request is from the edge function. It also adds a new file `route.tsx` that defines the edge function logic and a custom header `x-xlog-host` to identify edge requests. It also improves the note cover component by using a gradient of unique colors.

### <samp>🤖 Generated by Copilot at 92c5aeb</samp>

> _Oh we're the crew of the `xlog` ship and we sail the web so wide_
> _We make our notes look fancy with the Open Graph applied_
> _We heave and ho and pull the code, we push it to the edge_
> _And when we see the images, we give ourselves a pledge_

### WHY

CleanShot 2023-07-18 at 10 42 16@2x

### <samp>🤖 Generated by Copilot at 92c5aeb</samp>

*  Add a new edge function to generate dynamic Open Graph images for notes ([link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-3a3e69669e4740a29efc4ec38bedd1f8f6a480a4493718cb633dde469cd0deb6R1-R188))
*  Use the `api/og` endpoint to set the `og:image` meta tag for the note page if the request has a `x-xlog-host` header ([link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-81b6f6a35cbf3df1600ee46e2cb30e93f659c85f1cf33846c5f4ffad391323d3R2), [link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-81b6f6a35cbf3df1600ee46e2cb30e93f659c85f1cf33846c5f4ffad391323d3L41-R50), [link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-0626369278c89781505d2c04865ecce1302e979271bcfb471e029c19addc13a0R89))
*  Create a linear gradient background for the note cover using unique colors based on the note title ([link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-f9f1ef6cd01480d879e9aa3a4ebaca89a7b4ce3d945924213cf5a1a37f8e2315R16-R30), [link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-f9f1ef6cd01480d879e9aa3a4ebaca89a7b4ce3d945924213cf5a1a37f8e2315L21-R36))
*  Remove a trailing whitespace from the `package.json` file ([link](https://github.com/Crossbell-Box/xLog/pull/792/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L214-L213))

### CLAIM REWARDS

For first-time contributors, please leave your xLog address and Discord ID below to claim your rewards.

### CODE REVIEW

**Comment by DIYgod** [comment]
The App router already includes @vercel/og, so there is no need to install it. Refer to [this document](https://vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples)

**Comment by DIYgod** [comment]
It's quite a coincidence that I just did this as well. I have quit now and I can share with you the method I wrote for obtaining the avatar and title.

```ts
const result = await client
  .query(
    gql`
      query getNote($characterId: Int!, $noteId: Int!) {
        note(
          where: {
            note_characterId_noteId_unique: {
              characterId: $characterId
              noteId: $noteId
            }
          }
        ) {
          metadata {
            content
          }
          character {
            handle
            metadata {
              content
            }
          }
        }
      }
    `,
    {
      characterId: parseInt(characterId),
      noteId: parseInt(noteId),
    },
  )
  .toPromise()
const title = result.data.note.metadata?.content?.title
const avatar = result.data.note.character.metadata?.content?.avatars?.[0]
```

**Comment by Innei** [comment]
The test endpoint is `/api/og?characterId=52055&noteId=286&site=innei-4525`.

I replaced `[slug]/layout.tsx` SEO `openGraph` images, but not testing. I am not sure it should be worked.

**Comment by Innei** [comment]
The Font file will load from a remote URL because every serverless function size is limited to 1MB.

File: src/app/api/og/route.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+export const revalidate = 60 * 60 * 24 // 24 hours
+export const GET = async (req: NextRequest) => {
+  try {
+    const fontData = await fontNormal
+
+    // api/og?characterId=52055&noteId=286&site=innei-4525
+    const noteId = req.nextUrl.searchParams.get("noteId")
+    const characterId = req.nextUrl.searchParams.get("characterId")
+    const siteId = req.nextUrl.searchParams.get("site")
```
Comment (DIYgod) [inline src/app/api/og/route.tsx:23]: This is not used

File: src/app/site/[site]/[slug]/page.tsx [hunk 2]
```diff
@@ -38,13 +39,15 @@ export async function generateMetadata({
...
+  const reqHost = headers().get("x-xlog-host")
   const title = `${page?.metadata?.content?.title} - ${
     site?.metadata?.content?.name || site?.handle
   }`
+
   const description = page?.metadata?.content?.summary
-  const siteImages =
-    site?.metadata?.content?.avatars?.[0] || `${SITE_URL}/assets/logo.svg`
+  const siteImages = reqHost
+    ? `${reqHost}/api/og?noteId=${page?.noteId}&characterId=${site?.characterId}`
```
Comment (DIYgod) [inline src/app/site/[site]/[slug]/page.tsx:49]: It can use `SITE_URL` directly

**Review by DIYgod — CHANGES_REQUESTED** [review]

**Comment by DIYgod** [comment]
Thanks so much!

**Pull request merged**